File: bot.go
package main

import (
	"log"
	"time"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api"
	"github.com/jasonlvhit/gocron"
)

// Configuration parameters
const (
	Token            = "6765987978:AAHuqASdsQPFUV2O28nbx9nbM0QbTwHKL4U"
	SourceChannelID  = -1002023444138 // Replace with your source channel ID
	TargetChannelID  = -1002059668597 // Replace with your target channel ID
	ForwardBatchSize = 6
	ForwardInterval  = 10 // in minutes
)

func main() {
	bot, err := tgbotapi.NewBotAPI(Token)
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("Authorized on account %s", bot.Self.UserName)

	// Fetch and forward messages from source channel
	fetchAndForwardMessages(bot)

	// Schedule forwarding every 10 minutes
	gocron.Every(ForwardInterval).Minutes().Do(fetchAndForwardMessages, bot)

	// Run the scheduler
	<-gocron.Start()

	// Keep the program running
	select {}
}

func fetchAndForwardMessages(bot *tgbotapi.BotAPI) {
	// Get all messages from source channel
	allMessages, err := getAllMessages(bot)
	if err != nil {
		log.Printf("Error fetching messages: %v", err)
		return
	}

	// Forward videos to target channel in batches
	for i := 0; i < len(allMessages); i += ForwardBatchSize {
		end := i + ForwardBatchSize
		if end > len(allMessages) {
			end = len(allMessages)
		}

		batch := allMessages[i:end]
		for _, message := range batch {
			if message.Video != nil {
				forwardMsg := tgbotapi.NewForward(TargetChannelID, SourceChannelID, message.MessageID)
				if _, err := bot.Send(forwardMsg); err != nil {
					log.Printf("Error forwarding message: %v", err)
				}
			}
		}

		// Sleep for 10 minutes
		time.Sleep(ForwardInterval * time.Minute)
	}
}

func getAllMessages(bot *tgbotapi.BotAPI) ([]tgbotapi.Message, error) {
	var allMessages []tgbotapi.Message
	offset := 0
	limit := 100
	for {
		messages, err := bot.GetChatHistory(tgbotapi.ChatConfig{ChatID: SourceChannelID, Limit: limit, Offset: offset})
		if err != nil {
			return nil, err
		}
		if len(messages.Messages) == 0 {
			break
		}
		allMessages = append(allMessages, messages.Messages...)
		offset += len(messages.Messages)
	}
	return allMessages, nil
}

File: bot.py
from pyrogram import Client
from pyrogram.errors import FloodWait
from config import API_ID, API_HASH, SOURCE_CHANNEL_ID, TARGET_CHANNEL_ID
import time

# Configuration parameters
FORWARD_BATCH_SIZE = 6
FORWARD_INTERVAL_MINUTES = 10

# Initialize Pyrogram client
app = Client("my_account", api_id=API_ID, api_hash=API_HASH)


@app.on_message()
def forward_videos(client, message):
    # Ensure the message is from the source channel and is a video
    if message.chat.id == SOURCE_CHANNEL_ID and message.video:
        try:
            # Forward the message to the target channel
            client.forward_messages(chat_id=TARGET_CHANNEL_ID, from_chat_id=SOURCE_CHANNEL_ID, message_ids=message.message_id)
            print("Video forwarded successfully.")
        except FloodWait as e:
            print(f"Sleeping for {e.x} seconds.")
            time.sleep(e.x)
        except Exception as e:
            print(f"Error: {e}")


def main():
    with app:
        # Fetch and forward messages from source channel when bot starts
        messages = app.get_chat_history(chat_id=SOURCE_CHANNEL_ID, limit=FORWARD_BATCH_SIZE)
        for message in messages:
            if message.video:
                app.send_message(chat_id=TARGET_CHANNEL_ID, text="Forwarding video...")
                app.forward_messages(chat_id=TARGET_CHANNEL_ID, from_chat_id=SOURCE_CHANNEL_ID, message_ids=message.message_id)
                time.sleep(1)  # Delay between forwarding each message to avoid flood

        # Schedule forwarding every 10 minutes
        while True:
            messages = app.get_chat_history(chat_id=SOURCE_CHANNEL_ID, limit=FORWARD_BATCH_SIZE)
            for message in messages:
                if message.video:
                    app.send_message(chat_id=TARGET_CHANNEL_ID, text="Forwarding video...")
                    app.forward_messages(chat_id=TARGET_CHANNEL_ID, from_chat_id=SOURCE_CHANNEL_ID, message_ids=message.message_id)
                    time.sleep(1)  # Delay between forwarding each message to avoid flood
            time.sleep(FORWARD_INTERVAL_MINUTES * 60)


if __name__ == '__main__':
    main()
